apps/strava/strava.star

```diff
@@ -0,0 +1,378 @@
+             ),
+         ]
+
+    return render.Root(
+        child = render.Column(
+            children = [
+                render.Row(
+                    cross_align = "center",
+                    children = display_header,
+                ),
+                render.Row(
+                    cross_align = "center",
+                    children = [
+                        render.Image(src = SPORT_ICON),
+                        render.Text(" %s " % humanize.comma(float(stats.get('count', 0)))),
+                        render.Text(actu, font="tb-8"),
+                    ],
+                ),
+                render.Row(
+                    cross_align = "center",
+                    children = [
+                        render.Image(src = DISTANCE_ICON),
+                        render.Text(" %s " % humanize.comma(float(stats.get('distance', 0)))),
+                        render.Text(distu, font="tb-8"),
+                    ],
+                ),
+                render.Row(
+                    cross_align = "center",
+                    children = third_stat,
+                ),
+            ],
+        ),
+    )
+
+
+def meters_to_mi(m):
+    return m * 0.00062137
+
+def meters_to_km(m):
+    return m / 1000
+
+def meters_to_ft(m):
+    return m * 3.280839895
+
+def round(num, precision):
+    return math.round(num * math.pow(10, precision)) / math.pow(10, precision)
+
+def format_duration(d):
+    m = int(d.minutes)
+    s = str(int((d.minutes - m) * 60))
+    m = str(m)
+    if len(s) == 1:
+        s = '0' + s
+    return '%s:%s' % (m, s)
+
+def oauth_handler(params):
+    params = json.decode(params)
+    auth_code = params.get('code')
+    return auth_code
+
+def get_access_token(access_code, secret):
+    params = dict(
+        code=access_code,
+        client_secret=secret[:-1],
+        grant_type="authorization_code",
+        client_id=CLIENT_ID,
+    )
+    query_params = '&'.join(['%s=%s' % (k, v) for k, v in params.items()])
+    print("https://www.strava.com/api/v3/oauth/token?%s" % query_params)
+    res = http.post(
+        url = "https://www.strava.com/api/v3/oauth/token?%s" % query_params,
+        headers = {
+          "Accept": "application/json",
+        },
+        form_encoding = "application/x-www-form-urlencoded",
+    )
+    if res.status_code != 200:
+        fail("token request failed with status code: %d - %s" %
+             (res.status_code, res.body()))
+
+    token_params = res.json()
+    refresh_token = token_params["refresh_token"]
+    access_token = token_params["access_token"]
+    athlete = int(float(token_params['athlete']['id']))
+
+    cache.set(GLOBAL_CACHE_PREFIX + 'athlete_id', str(athlete), ttl_seconds=CACHE_TTL)
+    cache.set(GLOBAL_CACHE_PREFIX + "access_token", access_token, ttl_seconds = int(token_params["expires_in"] - 30))
+    cache.set(GLOBAL_CACHE_PREFIX + "refresh_token", refresh_token, ttl_seconds = int(token_params["expires_in"] - 30))
```

Comment:
It's my understanding that the cache is global across all users. Are you sure you can store these tokens in the global cache? Wouldn't that mean that this app would only work for the first user who ever authenticated?

Reply:
Is that true? I can include the athlete ID in the key name if that's the case but would appreciate if someone could confirm that the cache is not local to the device.

Reply:
This was my understanding after reading the docs here (https://github.com/tidbyt/pixlet/blob/main/docs/authoring_apps.md):

> Make sure to create cache keys that are unique for the type of information you are caching. Each app has its own cache, but that cache is shared among every user. Two installations of the same app by two different users will share the same cache.

Reply:
Ok, is there an "instance ID" or something that I can include so people can have multiple Strava views on one device?

Reply:
I just asked the same question in my pull request. :) I was checking out your code to see how you were doing OAuth, as I'm not quite sure how to test it.

It's really clever how you put the token exchange in the render function. I could try that to test that part of my code.

Can I ask, how did you test the initial OAuth request? Did you just have it redirect the response to a server you own? Thanks!

Reply:
Thanks! I was able to test the oauth exchange using pixlet serve, you can test your config that way as well. 

The one thing I couldn't get working locally was the secret encryption, I assume only apps on device have access to the private key?

Reply:
Oh, thanks for the advice! Let me try `pixlet serve` again.

Re: Secrets
Yep! (Same doc as above.)

> When you run pixlet locally, secret.decrypt will always return None. When your app runs in the Tidbyt cloud, secret.decrypt will return the string that you passed to pixlet encrypt.

Reply:
Got it, thanks!!